Fix incorrect references to "HttpClientFactory" (#17292)
dotnet/docs: docs/architecture/microservices/implement-resilient-applications/use-httpclientfactory-to-implement-resilient-http-requests.md

---
title: Use IHttpClientFactory to implement resilient HTTP requests
description: Learn how to use IHttpClientFactory, available since .NET Core 2.1, for creating `HttpClient` instances, making it easy for you to use it in your applications. 
ms.date: 08/08/2019
---
# Use IHttpClientFactory to implement resilient HTTP requests

<xref:System.Net.Http.IHttpClientFactory> is a contract implemented by `DefaultHttpClientFactory`, an opinionated factory, available since .NET Core 2.1, for creating <xref:System.Net.Http.HttpClient> instances to be used in your applications.

## Issues with the original HttpClient class available in .NET Core

The original and well-known <xref:System.Net.Http.HttpClient> class can be easily used, but in some cases, it isn't being properly used by many developers.

> As a first issue, while this class implements `IDisposable`, using it with the `using` statement is not the best choice because when you dispose an `HttpClient` object, the underlying socket is not immediately released and this can cause a serious issue named ‘sockets exhaustion’. For more information about this issue, see the [You're using HttpClient wrong and it's destabilizing your software](https://aspnetmonsters.com/2016/08/2016-08-27-httpclientwrong/) blog post.

Reviewer: These changes didn't go in?

> As a first issue, while this class implements `IDisposable`, using it with the `using` statement is not the best choice because when you dispose an `HttpClient` object, the underlying socket is not immediately released and this can cause a serious issue named ‘sockets exhaustion’. For more information about this issue, see the [You're using HttpClient wrong and it's destabilizing your software](https://aspnetmonsters.com/2016/08/2016-08-27-httpclientwrong/) blog post.

Author: Nope, they didn't.

I still don't get it 🙁. 

I work from the changed files tabs, because that's the only way to batch updates, but it seems like there are different sets of suggestions.

Creating another commit...

Therefore, `HttpClient` is intended to be instantiated once and reused throughout the life of an application. Instantiating an `HttpClient` class for every request will exhaust the number of sockets available under heavy loads. That issue will result in `SocketException` errors. Possible approaches to solve that problem are based on the creation of the `HttpClient` object as singleton or static, as explained in this [Microsoft article on HttpClient usage](../../../csharp/tutorials/console-webapiclient.md). This can be a good solution for short-lived console apps or similar that are run a few times a day.

> But there’s a second issue with `HttpClient` that you can have when you use it as singleton or static object in long-running processes, as could be some sort of server. In this case, a singleton or static `HttpClient` doesn't handle DNS changes, as explained in this [issue](https://github.com/dotnet/corefx/issues/11224) at the dotnet/corefx GitHub repository.

Author: I'm tweaking this one a bit for you to review. I wouldn't mention the "to improve performance" but focus on the "long-running" processes, b/c that's when the DNS refresh problem could be an issue.

> To address those mentioned issues and make the management of `HttpClient` instances easier, .NET Core 2.1 introduced the <xref:System.Net.Http.IHttpClientFactory> interface that's implemented by the `DefaultHttpClientFactory`, that can't be instantiated directly, that can also be used to implement resilient HTTP calls by integrating Polly with it.

Reviewer: ```suggestion
To address the issues mentioned above and to make `HttpClient` instances manageable, .NET Core 2.1 introduced the <xref:System.Net.Http.IHttpClientFactory> interface which can be used to configure and create HttpClient instances in an app. It also provides extensions for Polly-based middleware to take advantage of delegating handlers in HttpClient.
```

[Polly](http://www.thepollyproject.org/) is a transient-fault-handling library that helps developers add resiliency to their applications, by using some pre-defined policies in a fluent and thread-safe manner.

> ## What is the IHttpClientFactory

Reviewer: IMO, we are talking about the benefits here. The earlier para concludes with what IHttpClientFactory is.

> ## What is the IHttpClientFactory

Author: That's right!

> <xref:System.Net.Http.IHttpClientFactory> is an interface with the single method `CreateClient()` that should return an <xref:System.Net.Http.HttpClient>.

Reviewer: ```suggestion
```

Reviewer: Remove this sentence as it is only describing what the interface is.

> The current implementation is `DefaultHttpClientFactory`, that also implements <xref:ystem.Net.Http.IHttpMessageHandlerFactory>, with the following features:

Reviewer: ```suggestion
`IHttpClientFactory` offers the following benefits:
```

Author: Well... not really, it'd be the current implementation of...

Will update!

- Codify the concept of outgoing middleware via delegating handlers in `HttpClient` and implementing Polly-based middleware to take advantage of Polly’s policies for resiliency.
- `HttpClient` already has the concept of delegating handlers that could be linked together for outgoing HTTP requests. You can register HTTP clients into the factory and you can use a Polly handler to use Polly policies for Retry, CircuitBreakers, and so on.
- Manage the lifetime of <xref:System.Net.Http.HttpMessageHandler> to avoid the mentioned problems/issues that can occur when managing `HttpClient` lifetimes yourself.

> [!NOTE]
> The implementation of `IHttpClientFactory` (`DefaultHttpClientFactory`) is tightly tied to the dependency injection (DI) implementation in the `Microsoft.Extensions.DependencyInjection` NuGet package. For more information about using other dependency injection containers, see this [GitHub discussion](https://github.com/dotnet/extensions/issues/1345).

## Multiple ways to use IHttpClientFactory

There are several ways that you can use `IHttpClientFactory` in your application:

> - Use `IHttpClientFactory` Directly

Reviewer: ```suggestion
- Basic usage
```

## How to use Typed Clients with IHttpClientFactory

So, what's a "Typed Client"? It's just an `HttpClient` that's pre-configured for some specific use. This configuration can include specific values such as the base server, HTTP headers or time outs.

The following diagram shows how Typed Clients are used with `IHttpClientFactory`:

![Diagram showing how typed clients are used with IHttpClientFactory.](./media/use-httpclientfactory-to-implement-resilient-http-requests/client-application-code.png)

**Figure 8-4**. Using `IHttpClientFactory` with Typed Client classes.

In the above image, a `ClientService` (used by a controller or client code) uses an `HttpClient` created by the registered `IHttpClientFactory`. This factory assigns an `HttpMessageHandler` from a pool to the `HttpClient`. The `HttpClient` can be configured with Polly's policies when registering the `IHttpClientFactory` in the DI container with the extension method <xref:Microsoft.Extensions.DependencyInjection.HttpClientFactoryServiceCollectionExtensions.AddHttpClient*>.

> To configure the above structure, install the [`Microsoft.Extensions.Http`](https://www.nuget.org/packages/Microsoft.Extensions.Http/) NuGet package that includes the `AddHttpClient()` extension method for `IServiceCollection`. This extension method registers the `DefaultHttpClientFactory` as a singleton, as the implementation of `IHttpClientFactory`. It defines a transient configuration for the <xref:Microsoft.Extensions.Http.HttpMessageHandlerBuilder>. This message handler (<xref:System.Net.Http.HttpMessageHandler> object), taken from a pool, is used by the `HttpClient` returned from the factory.

Reviewer: revert this changes. previous one reads better.

Author: ok
